Closes #484 roclet: Implement rdx roclet

bundfussr opened the pull request:
Thank you for your Pull Request! We have developed this task checklist from the [Development Process Guide](https://pharmaverse.github.io/admiral/dev/CONTRIBUTING.html) to help with the final steps of the process. Completing the below tasks helps to ensure our reviewers can maximize their time on your code as well as making sure the admiral codebase remains robust and consistent.   

Please check off each taskbox as an acknowledgment that you completed the task or check off that it is not relevant to your Pull Request. This checklist is part of the Github Action workflows and the Pull Request will not be merged into the `main` branch until you have checked off each task.

- [x] Place Closes #<insert_issue_number> into the beginning of your Pull Request Title (Use Edit button in top-right if you need to update)
- [x] Code is formatted according to the [tidyverse style guide](https://style.tidyverse.org/). Run `styler::style_file()` to style R and Rmd files
- [x] Updated relevant unit tests or have written new unit tests, which should consider realistic data scenarios and edge cases, e.g. empty datasets, errors, boundary cases etc. - See [Unit Test Guide](https://pharmaverse.github.io/admiraldev/articles/unit_test_guidance.html#tests-should-be-robust-to-cover-realistic-data-scenarios)
- [x] If you removed/replaced any function and/or function parameters, did you fully follow the [deprecation guidance](https://pharmaverse.github.io/admiraldev/articles/programming_strategy.html#deprecation)?
- [x] Update to all relevant roxygen headers and examples, including keywords and families. Refer to the [categorization of functions](https://pharmaverse.github.io/admiraldev/articles/programming_strategy.html#categorization-of-functions) to tag appropriate keyword/family.
- [x] Run `devtools::document()` so all `.Rd` files in the `man` folder and the `NAMESPACE` file in the project root are updated appropriately
- [x] Address any updates needed for vignettes and/or templates
- [x] Update `NEWS.md` under the header `# admiral (development version)` if the changes pertain to a user-facing function (i.e. it has an `@export` tag) or documentation aimed at users (rather than developers)
- [x] Build admiral site `pkgdown::build_site()` and check that all affected examples are displayed correctly and that all new functions occur on the "[Reference](https://pharmaverse.github.io/admiraldev/reference/index.html)" page. 
- [x] Address or fix all lintr warnings and errors - `lintr::lint_package()`
- [x] Run `R CMD check` locally and address all errors and warnings - `devtools::check()`
- [x] Link the issue in the Development Section on the right hand side.
- [x] Address all merge conflicts and resolve appropriately
- [x] Pat yourself on the back for a job well done! Much love to your accomplishment!

manciniedoardo commented:
@bundfussr as discussed in today's call, if you can get to this PR within the next week/10 days, then this will unlock some of the documentation work on the main admiral core website :)

manciniedoardo commented on R/roclet_rdx.R:
```diff
@@ -0,0 +1,585 @@
+#'
+#' @keywords documentation
+#'
+#' @export
```
It would be cool to showcase the extra tags in an example section here, but not sure if that even makes sense or is possible?

bundfussr replied:
Yes, that would be cool. But I'm not sure how to do it in a well readable way. We could add an example like:
```
> roxygen2::roc_proc_text(
+     rdx_roclet(),
+     c(
+         "#' Testing",
+         "#' @param x An argument",
+         "#' @permitted A number",
+         "test_fun <- function(x) 42"
+     ))
$test_fun.Rd
% Generated by roxygen2: do not edit by hand
% Please edit documentation in ./<text>
\name{test_fun}
\alias{test_fun}
\title{Testing}
\usage{
test_fun(x)
}
\arguments{
\item{x}{An argument

\describe{
\item{Permitted values}{A number}
\item{Default value}{none}
}}
}
\description{
Testing
}
```
What do you think?

manciniedoardo replied:
Yes, I was envisaging something like that - could you pls put in the PR so we can render and review?

bundfussr replied:
I've extended the example a bit such that all custom tags occur.

bms63 replied:
What about just making a vignette for this and linking to that in our function documentation and in the programming strategy.  feels like we need more room!

manciniedoardo reviewed (commented):
Thanks @bundfussr for this 😄 I left some thoughts after reviewing the documentation, here are some other comments:

- I noticed the documentation for the functions in `R/roclet_rdx.R` is not indexed in the reference page, but I think there would be use in seeing it there, especially under its own category?
- I wonder if we should be a little bit louder about the fact we have/use custom roclets, at least for complex functions. In the programming strategy it's currently a bit buried in the Function Header (Documentation) section, so maybe we should say it right at the top of that section before jumping into the details - something like: "Every function that is exported from the package must have an accompanying header that should be formatted according to the [roxygen2](https://roxygen2.r-lib.org/) convention". We also implement custom roclets to enhance our documentation and examples for more complex functions - see rdx_roclet and further down for more details."

bundfussr commented:
> * I noticed the documentation for the functions in `R/roclet_rdx.R` is not indexed in the reference page, but I think there would be use in seeing it there, especially under its own category?

There is a new "Documentation Functions" section on the reference page.

bundfussr commented on vignettes/programming_strategy.Rmd:
```diff
@@ -395,23 +401,38 @@ An example is given below:
-If a dataset is created in the example, it should be done so using the function `tribble()` (specify `library(dplyr)` before calling this function).
-If other functions are called in the example, please specify `library(pkg_name)` then refer to the respective function `fun()` as opposed to the preferred `pkg_name::fun()` notation as specified in [Unit Test Guidance](unit_test_guidance.html#set-up-the-test-script).
-Make sure to align columns as this ensures quick code readability. 
+* `@examples` or `@caption`, `@info`, `@code`: Fully self-contained examples of how to use the function.
```
Added below as suggested

bundfussr commented on vignettes/programming_strategy.Rmd:
```diff
@@ -395,23 +401,38 @@ An example is given below:
+Guidance](unit_test_guidance.html#set-up-the-test-script). Make sure to align
+columns as this ensures quick code readability.
+
+    The `@examples` tag should be used for simple functions which requires only
```
fixed

bundfussr commented on vignettes/programming_strategy.Rmd:
```diff
@@ -395,23 +401,38 @@ An example is given below:
+columns as this ensures quick code readability.
+
+    The `@examples` tag should be used for simple functions which requires only
+    a few examples and no explanation. For more complex functions, the
```
Added

bundfussr commented:
> * I wonder if we should be a little bit louder about the fact we have/use custom roclets, at least for complex functions. In the programming strategy it's currently a bit buried in the Function Header (Documentation) section, so maybe we should say it right at the top of that section before jumping into the details - something like: "Every function that is exported from the package must have an accompanying header that should be formatted according to the [roxygen2](https://roxygen2.r-lib.org/) convention". We also implement custom roclets to enhance our documentation and examples for more complex functions - see rdx_roclet and further down for more details."

Yes, updated as suggested.

I have also added some more details regarding examples for complex functions.

bms63 commented:
Apologies @bundfussr I have this on my list for tomorrow.

manciniedoardo commented:
> > * I noticed the documentation for the functions in `R/roclet_rdx.R` is not indexed in the reference page, but I think there would be use in seeing it there, especially under its own category?
> 
> There is a new "Documentation Functions" section on the reference page.

Sorry, I'd missed it!

bms63 commented on NEWS.md:
```diff
@@ -2,6 +2,11 @@
+- The `rdx_roclet()` extends the standard `{roxygen2}` roclet by the `@permitted`
+and `@default` tags to specify the permitted and default values and the
+`@caption`, `@info`, and `@code` tags for examples with a caption and a
+description.
```
```suggestion
description. (#484)
```

bundfussr replied:
Yes

bms63 commented on R/quo.R:
```diff
@@ -105,17 +105,16 @@ replace_symbol_in_expr <- function(expression,
 #'
-#'   *Permitted Values*: list of variables or `desc(<variable>)` function calls
-#'   created by `exprs()`, e.g., `exprs(ADT, desc(AVAL))`
+#' @permitted list of variables or `desc(<variable>)` function calls created by
```
```suggestion
#' @permitted list of variables or `dplyr::desc(<variable>)` function calls created by
```
While I am not a prefix person - wondering if this might be helpful for users or even devs is we put this in there??

bundfussr replied:
I wouldn't add the prefix because it suggests that it is required. But it isn't. You don't even need the prefix if `dplyr` is not loaded.

bundfussr commented:
> Do we need to align roxygennote with roxygen package version??

Isn't it updated automatically?

Seems to be the latest version.

bms63 commented on R/roclet_rdx.R:
Shall we link out to roxygen https://roxygen2.r-lib.org/articles/extending.html as I think many developers won't be familiar (like me!!)

bundfussr replied:
Yes, good idea!

bms63 commented on man/demo_fun.Rd:
yeah!  feels like we could take advantage of this being generated to play around with this for an example

bms63 replied:
actually...ummm shouldn't more come through to the Rd file?

bundfussr replied:
I have updated the header of `demo_fun()` such that it is in line with the text used in the example.

manciniedoardo replied:
Looks great! I'd also link `demo_fun` in the get started section and also mention in the Programming Startegy somewhere.

bundfussr replied:
Do you mean the source file, the rendered documentation or both?

manciniedoardo replied:
I'd showcase the rendered documentation page for demo_fun, as that then links to the rdx function

bundfussr replied:
Links added

bundfussr commented on README.md:
```diff
@@ -4,23 +4,23 @@
 Utility Functions and Development Tools for the Admiral Package Family
 
-<!-- badges: start -->
+<!-- start badges -->
```
@bms63 , this makes the badges appear on both the main and the dev site:

![image](https://github.com/user-attachments/assets/f84dc1e4-4d7b-4b04-9692-d0ac90002087)

What do you think?

bms63 replied:
looks great to me!! I will adjust!

bundfussr commented on R/assertions.R:
```diff
@@ -5,7 +5,7 @@
 #' @param arg A function argument to be checked
 #' @param required_vars A list of variables created using `exprs()`
-#' @param check_is_grouped Throw an error is `dataset` is grouped? Defaults to `TRUE`.
+#' @param check_is_grouped Throw an error is `dataset` is grouped?
```
Updated

bundfussr commented on R/expect_dfs_equal.R:
```diff
@@ -4,6 +4,11 @@
+#'
+#' @permitted A dataset
+#'
+#'   Which means a data.frame or tibble.
```
I think this would be misleading as every tibble is also a data.frame.

I have updated it and put everything in one line. Most likely it is a left-over from testing.

bundfussr commented on R/roclet_rdx.R:
```diff
@@ -0,0 +1,617 @@
+#' Demo function for `rdx_roclet()` example
+#'
+#' @keywords internal
+demo_fun <- function(x, number = 1, letter = "a") 42
```
Copilot suggests the text in gray 😄:
![image](https://github.com/user-attachments/assets/89e92510-740b-4b5e-a8f5-dcc2cd77fc1e)

bms63 replied:
yes!!!

manciniedoardo reviewed (approved):
Love it!!

bms63 commented:
I took it for a spin and made a silly function with examples.

Feels very nice!!  Nice work @bundfussr feels like this will be a significant milestone for admiral!!

![image](https://github.com/user-attachments/assets/4b7ebd2d-9d1b-4923-be30-a687d3cf3324)

![image](https://github.com/user-attachments/assets/1441d0a6-1bbd-435f-a5d8-de29102fff48)
